### PR TITLE
Paged attention

### PR DESCRIPTION
**Pull request by liangan1**

### CODE REVIEW

File: test/kernel/test_paged_attention.py [hunk 1]
```diff
@@ -0,0 +1,206 @@
+import torch
+import unittest
+import random
+from typing import List, Optional, Tuple
+from itertools import product
+import torchao
+from torchao.kv_cache import PagedAttentionCache, PagedTensor
+
+class NiaveCache:
```
Comment (jgong5) [inline test/kernel/test_paged_attention.py]:
```suggestion
class NaiveCache:
```

File: torchao/kv_cache.py [hunk 2]
```diff
@@ -0,0 +1,415 @@
...
+HANDLED_FUNCTIONS = {}
+
+
+class PagedTensor(object):
+    def __init__(
+        self,
+        cache: torch.Tensor,
+        block_tables: torch.Tensor,
+        context_lens: torch.Tensor,
+    ):
```
Comment (jgong5) [inline torchao/kv_cache.py]: Consider adding document on these args? Also, would be helpful to note what are owned by the object and what are shared? I guess `cache` is shared among multiple `PagedTensor`s? What are the shapes for these tensors?

File: torchao/kv_cache.py [hunk 3]
```diff
@@ -0,0 +1,415 @@
...
+        self.block_tables = block_tables
+        self.cache = cache
+        self.context_lens = context_lens
```
Comment (jgong5) [inline torchao/kv_cache.py]: Not sure if it is good or general enough to incorporate "context length" into the semantics of a `PagedTensor`. The context length sounds like an app-level concept, not a general tensor-level concept?

File: torchao/kv_cache.py [hunk 4]
```diff
@@ -0,0 +1,415 @@
...
+
+@implements(torch.nn.functional.scaled_dot_product_attention)
+def scaled_dot_product_attention(
+    input, key_tensor, value_tensor, attn_mask=None, scale=None
+):
+    query = input
+    key_cache = key_tensor.cache
+    value_cache = value_tensor.cache
+    num_kv_head = key_cache.size(1)
+    num_queries_per_kv = query.size(1) // num_kv_head
```
Comment (jgong5) [inline torchao/kv_cache.py]: Should we add an assertion here to make sure `query.size(1) % num_kv_head == 0`?

File: torchao/kv_cache.py [hunk 5]
```diff
@@ -0,0 +1,415 @@
...
+        torch.arange(num_kv_head, dtype=torch.int32, device="cpu"), num_queries_per_kv
+    )
+    context_lens = key_tensor.context_lens
+    output = torch.empty_like(query)
+    torch.ops.torchao.paged_attention(
+        output,
+        query,
+        key_cache,
+        value_cache,
+        head_mapping,
```
Comment (jgong5) [inline torchao/kv_cache.py]: Can we remove this `head_mapping` and move it into the implementation? Assume we always do the even mapping here.

File: torchao/csrc/cpu/paged_attention_kernel.cpp [hunk 6]
```diff
@@ -0,0 +1,535 @@
...
+  auto exp_sum_strideN = exp_sum.stride(0);
+  auto exp_sum_strideH = exp_sum.stride(1);
+  auto tmp_out_strideN = tmp_out.stride(0);
+  auto tmp_out_strideH = tmp_out.stride(1);
+  auto tmp_out_strideS = tmp_out.stride(2);
+#pragma omp parallel for collapse(3) schedule(static, 1)
+  for (auto partition_id = 0; partition_id < max_num_partitions;
+       partition_id++) {
+    for (auto head_id = 0; head_id < num_heads; head_id++) {
+      for (auto seq_id = 0; seq_id < num_seqs; seq_id++) {
```
Comment (jgong5) [inline torchao/csrc/cpu/paged_attention_kernel.cpp]: Is this loop order the most optimal one?

File: torchao/csrc/cpu/paged_attention_kernel.cpp [hunk 7]
```diff
@@ -0,0 +1,535 @@
...
+        auto kv_head_id = head_id / kv_head_group_size;
+        auto q_ptr_start = query_ptr + seq_id * q_strideN + head_id * q_strideH;
+        auto max_logits_offset = seq_id * max_logits_strideN +
+                                 head_id * max_logits_strideH + partition_id;
+        auto exp_sum_offset =
+            seq_id * exp_sum_strideN + head_id * exp_sum_strideH + partition_id;
+        auto tmp_out_start = tmp_out_ptr + seq_id * tmp_out_strideN +
+                             head_id * tmp_out_strideH +
+                             partition_id * tmp_out_strideS;
+        accum_t logits[PARTITION_SIZE] __attribute__((aligned(64))) = {0};
```
Comment (jgong5) [inline torchao/csrc/cpu/paged_attention_kernel.cpp]: use `std::alignas`

File: torchao/csrc/cpu/paged_attention_kernel.cpp [hunk 8]
```diff
@@ -0,0 +1,535 @@
...
+
+  int64_t mStrideB = (has_attn_mask && attn_mask.value().size(0) > 1)
+                         ? attn_mask.value().stride(0)
+                         : 0;
+  int64_t mStrideH = (has_attn_mask && attn_mask.value().size(1) > 1)
+                         ? attn_mask.value().stride(1)
+                         : 0;
+  int64_t mStrideM = has_attn_mask ? attn_mask.value().stride(2) : 0;
+
+  auto max_num_partitions =
```
Comment (jgong5) [inline torchao/csrc/cpu/paged_attention_kernel.cpp]: nit: The name `partition` sounds too general. Suggest to specify it is for sequence, e.g., `max_num_seq_partitions`. Same comments for other related names.

File: torchao/csrc/cpu/paged_attention_kernel.cpp [hunk 9]
```diff
@@ -0,0 +1,535 @@
...
+ * attention.
+ *
+ * This function computes the attention weights and applies the attention
+ * mechanism to obtain the final output. It takes in tensors representing the
+ * query, key cache, value cache, head mapping, scale, block tables, context
+ * lengths, block size The
+ * output tensor is updated with the computed attention values.
+ *
+ * @param out           Output tensor [num_seqs, 1, num_heads, head_size].
+ * @param query         Query tensor [num_seqs, 1, num_heads, head_size].
```
Comment (jgong5) [inline torchao/csrc/cpu/paged_attention_kernel.cpp]: Please add runtime assertion in the code to make sure the query has seq length 1 here. BTW, can we extend the implementation to support query seq length > 1 which can benefit chunked prefill and multi-turn conversation cases?

File: torchao/kv_cache.py [hunk 10]
```diff
@@ -0,0 +1,415 @@
...
+    query = input
+    key_cache = key_tensor.cache
+    value_cache = value_tensor.cache
+    num_kv_head = key_cache.size(1)
+    num_queries_per_kv = query.size(1) // num_kv_head
+    block_size = key_cache.size(2)
+    block_tables = key_tensor.block_tables
+    head_mapping = torch.repeat_interleave(
+        torch.arange(num_kv_head, dtype=torch.int32, device="cpu"), num_queries_per_kv
+    )
```
Comment (jgong5) [inline torchao/kv_cache.py]: Can we do this inside the `paged_attention` c++ kernel so that we don't need to pass this `head_mapping` arg to it? This simplifies the kernel interface.

File: torchao/csrc/cpu/paged_attention_kernel.cpp [hunk 11]
```diff
@@ -0,0 +1,535 @@
...
+  auto num_heads = query.size(1);
+  auto head_size = query.size(3);
+  auto num_kv_heads = key_cache.size(1);
+  auto max_num_blocks_per_seq = block_tables.size(1);
+  auto kv_head_group_size = num_heads / num_kv_heads;
+  bool has_attn_mask = attn_mask.has_value() && attn_mask.value().numel();
+  if (has_attn_mask) {
+    attn_mask.value() = attn_mask.value().to(at::kFloat);
+    reshape_attn_mask_to_4d(attn_mask.value(), num_seqs, num_heads, q_len,
+                            attn_mask.value().size(-1));
```
Comment (jgong5) [inline torchao/csrc/cpu/paged_attention_kernel.cpp]: Does this convert the attn_mask to 4D or just view it as 4D? Since we are working on raw pointers, perhaps we don't need to expand it as 4D view here?

File: torchao/csrc/cpu/paged_attention_kernel.cpp [hunk 12]
```diff
@@ -0,0 +1,535 @@
...
+                                           head_size);
+            logits_position++;
+          }
+        }
+        // 2) calculate the max and exp_sum for this partition
+        auto partition_max = -std::numeric_limits<accum_t>::infinity();
+        if (has_attn_mask) {
+          _scale_attn_mask_fusion_kernel<accum_t, accum_t>(
+              logits,
+              attn_mask_ptr + seq_id * mStrideB + head_id * mStrideH +
```
Comment (jgong5) [inline torchao/csrc/cpu/paged_attention_kernel.cpp:366]: I guess we need to carefully handle the case where the size is 1 in some dim of the mask here.

**Review by jgong5 — CHANGES_REQUESTED** [review]
